firedrake/slate/slate.py

```diff
@@ -589,41 +592,23 @@ def __init__(self, tensor, coefficient):
             "Argument function space must be the same as the "
             "coefficient function space."
         )
-        super(Action, self).__init__()
-        self.operands = tensor, coefficient
+        super(Action, self).__init__(tensor,
+                                     assembled_coeffs=(coefficient,))
```

Comment:
I think the superclass shouldn't know about this extra coefficient junk.  But rather than you should override `coefficients` here:

```python
def coefficients(self):
    return tuple(OrderedDict.fromkeys(super(Action, self).coefficients() + (self.actee, )))
```

Similarly with `ufl_domains`.  If you're worried about code reuse in `join_domains` and similar, make free functions:

`merge_coefficients` and `join_domains` and just call them with the appropriate arguments in the instance methods.